File: euler4.rb
# Build a list of all the palindromes. Check if each one is the product of
# two three-digit numbers by dividing each by every three digit number.
# O( n^2 ) for n = 10^digits
# Actually ends up checking 93 palindromes for 3 digits.
# Takes under 1 second for 4 digits number on macbook pro.

def maxpalindrome(digits)
  # Maximum first half of palindrome number is the first three digits of the
  # square of the maximum n-digit number.
  counter=0
  maxhalfpal = (10**digits-1)**2 / 10**digits
  i=maxhalfpal
  minfactor = 10**(digits-1)
  maxfactor = 10**digits-1
  while i>minfactor
    curr = i*10**digits
    for j in 0..digits-1
      adder = (i/10**j % 10) * 10**(digits-j-1)
      curr += adder
    end
    curr = Float(curr)
    counter+=1
    for j in minfactor..maxfactor
      res = curr / Float(j)
      if res % 1 == 0 and res >= minfactor and res <= maxfactor
        p counter
        return Integer(curr)
      end
    end
    i-=1
  end
end

digits = 3  # Should be 906609
#digits = 4
#digits = 2
p maxpalindrome(digits)


# Build a list of all the 3-digit products. Sort them. Check for palindromes
# from top to bottom. O( n^2 log n )
# Takes 15 seconds for 4 digits on macbook pro.
def maxpalindrome2(digits)
  minfactor = 10**(digits-1)
  maxfactor = 10**digits-1
  products = []
  for i in minfactor..maxfactor
    for j in i..maxfactor
      products.push(i*j)
    end
  end
  products = products.sort!.reverse!
  products.each do |i|
    istr = i.to_s
    if istr == istr.reverse
      return i
    end
  end
end
#p maxpalindrome2(digits)


# First idea:
# Monontically decreasing product of two 3-digit numbers, then check for
# palindrome. This doesn't work because it's hard to get a monotonic order of
# products.

File: golang/euler4.go
// Find the largest palindrome made from the product of two 3-digit numbers.

package main

import (
	"fmt"
)
import "./numberutil"

func max_palindrome(digits int) int {
	// Maximum first half is the first three digits of the maximum factor squared.
	max_first_half := numberutil.Pow_int_to_int(numberutil.Pow_10_to_int(digits)-1, 2) / numberutil.Pow_10_to_int(digits)
	min_factor := numberutil.Pow_10_to_int(digits - 1)
	max_factor := numberutil.Pow_10_to_int(digits) - 1
	for prefix := max_first_half; prefix >= min_factor; prefix-- {
		number := prefix * numberutil.Pow_10_to_int(digits)
		for i := 0; i < digits; i++ {
			number += prefix / numberutil.Pow_10_to_int(i) % 10 * numberutil.Pow_10_to_int(digits-i-1)
		}
		for factor := min_factor; factor <= max_factor; factor++ {
			if number%factor == 0 {
				quotient := number / factor
				if quotient >= min_factor && quotient <= max_factor {
					return number
				}
			}
		}
	}
	return -1
}

func main() {
	result := max_palindrome(3)
	fmt.Println(result)
}
